Bottom: c2719813e1012a3aca4fb8c2c40ef5234394c51f
Top:    d4514a28fa3d5a69d4d6926c2c8b2ae7c340103a
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2012-12-22 15:40:22 +0000

Refresh of add-split-colouring-for-find-a

---

diff --git a/src/action.c b/src/action.c
index 9be97eb..74370d5 100644
--- a/src/action.c
+++ b/src/action.c
@@ -341,6 +341,7 @@ static FunctionType Functions[] = {
   {"ClearAndRedraw", F_ClearAndRedraw},
   {"ClearList", F_ClearList},
   {"Close", F_Close},
+  {"Found", F_Found},
   {"Connection", F_Connection},
   {"Convert", F_Convert},
   {"Copy", F_Copy},
@@ -2290,55 +2291,34 @@ ActionConnection (int argc, char **argv, Coord x, Coord y)
 	case F_Find:
 	  {
 	    gui->get_coords (_("Click on a connection"), &x, &y);
-	    LookupConnection (x, y, true, 1, FOUNDFLAG, false);
-	    LookupConnection (x, y, true, 1, RATFOUNDFLAG, true);
+	    LookupConnection (x, y, true, 1, FOUNDFLAG, true);
+	    LookupConnection (x, y, true, 1, CONNECTEDFLAG, false);
 	    break;
 	  }
 
 	case F_ResetLinesAndPolygons:
-	  {
-	    bool change = false;
-
-	    change = ResetFoundLinesAndPolygons (true, RATFOUNDFLAG) || change;
-	    change = ResetFoundLinesAndPolygons (true, FOUNDFLAG) || change;
-
-	    if (change)
-	      {
-	        IncrementUndoSerialNumber ();
-	        Draw ();
-	      }
-	    break;
-	  }
+	  if (ResetFoundLinesAndPolygons (true, FOUNDFLAG | CONNECTEDFLAG))
+	    {
+	      IncrementUndoSerialNumber ();
+	      Draw ();
+	    }
+	  break;
 
 	case F_ResetPinsViasAndPads:
-	  {
-	    bool change = false;
-
-	    change = ResetFoundPinsViasAndPads (true, RATFOUNDFLAG) || change;
-	    change = ResetFoundPinsViasAndPads (true, FOUNDFLAG) || change;
-
-	    if (change)
-	      {
-	        IncrementUndoSerialNumber ();
-	        Draw ();
-	      }
-	    break;
-	  }
+	  if (ResetFoundPinsViasAndPads (true, FOUNDFLAG | CONNECTEDFLAG))
+	    {
+	      IncrementUndoSerialNumber ();
+	      Draw ();
+	    }
+	  break;
 
 	case F_Reset:
-	  {
-	    bool change = false;
-
-	    change = ResetConnections (true, RATFOUNDFLAG) || change;
-	    change = ResetConnections (true, FOUNDFLAG) || change;
-
-	    if (change)
-	      {
-	        IncrementUndoSerialNumber ();
-	        Draw ();
-	      }
-	    break;
-	  }
+	  if (ResetConnections (true, FOUNDFLAG | CONNECTEDFLAG))
+	    {
+	      IncrementUndoSerialNumber ();
+	      Draw ();
+	    }
+	  break;
 	}
       return 0;
     }
@@ -5337,9 +5317,12 @@ Selects all objects in a rectangle indicated by the cursor.
 @item All
 Selects all objects on the board.
 
-@item Connection
+@item Found
 Selects all connections with the ``found'' flag set.
 
+@item Connection
+Selects all connections with the ``connected'' flag set.
+
 @item Convert
 Converts the selected objects to an element.  This uses the highest
 numbered paste buffer.
@@ -5444,7 +5427,17 @@ ActionSelect (int argc, char **argv, Coord x, Coord y)
 	    break;
 	  }
 
-	  /* all found connections */
+	  /* all logical connections */
+	case F_Found:
+	  if (SelectFound (true))
+	    {
+              Draw ();
+	      IncrementUndoSerialNumber ();
+	      SetChangedFlag (true);
+	    }
+	  break;
+
+	  /* all physical connections */
 	case F_Connection:
 	  if (SelectConnection (true))
 	    {
@@ -5623,7 +5616,17 @@ ActionUnselect (int argc, char **argv, Coord x, Coord y)
 	    break;
 	  }
 
-	  /* all found connections */
+	  /* all logical connections */
+	case F_Found:
+	  if (SelectFound (false))
+	    {
+              Draw ();
+	      IncrementUndoSerialNumber ();
+	      SetChangedFlag (true);
+	    }
+	  break;
+
+	  /* all physical connections */
 	case F_Connection:
 	  if (SelectConnection (false))
 	    {
diff --git a/src/const.h b/src/const.h
index c6bb881..98cf03c 100644
--- a/src/const.h
+++ b/src/const.h
@@ -214,7 +214,7 @@ pad.  Primarily used for pads used as fiducials.
 #define EDGE2FLAG                 0x4000	/* Padr.Point2 is closer to outside edge */
 						/* also pinout text for pins is vertical */
 #define VISITFLAG		  0x8000	/* marker to avoid re-visiting an object */
-#define	RATFOUNDFLAG		0x010000	/* used by 'FindConnection()' */
+#define	CONNECTEDFLAG		0x010000	/* flag like FOUND flag, but used to identify physically connected objects (not rats) */
 /* ---------------------------------------------------------------------------
  * PCB flags
  */
diff --git a/src/create.c b/src/create.c
index 7919b43..e167dc8 100644
--- a/src/create.c
+++ b/src/create.c
@@ -105,6 +105,7 @@ pcb_colors_from_settings (PCBType *ptr)
 
   /* copy default settings */
   ptr->ConnectedColor = Settings.ConnectedColor;
+  ptr->FoundColor = Settings.FoundColor;
   ptr->ElementColor = Settings.ElementColor;
   ptr->RatColor = Settings.RatColor;
   ptr->InvisibleObjectsColor = Settings.InvisibleObjectsColor;
diff --git a/src/draw.c b/src/draw.c
index 2b56791..61e6d99 100644
--- a/src/draw.c
+++ b/src/draw.c
@@ -85,17 +85,16 @@ static void DrawEMark (ElementType *, Coord, Coord, bool);
 static void DrawRats (const BoxType *);
 
 static void
-set_object_color (AnyObjectType *obj,
-                  char *warn_color, char *selected_color,
-                  char *found_color, char *normal_color)
+set_object_color (AnyObjectType *obj, char *warn_color, char *selected_color,
+                  char *connected_color, char *found_color, char *normal_color)
 {
   char *color;
 
-  if      (warn_color     != NULL && TEST_FLAG (WARNFLAG,     obj)) color = warn_color;
-  else if (selected_color != NULL && TEST_FLAG (SELECTEDFLAG, obj)) color = selected_color;
-  else if (found_color    != NULL && TEST_FLAG (FOUNDFLAG,    obj)) color = found_color;
-  else if (found_color    != NULL && TEST_FLAG (RATFOUNDFLAG, obj)) color = "#FF00FF";
-  else                                                              color = normal_color;
+  if      (warn_color      != NULL && TEST_FLAG (WARNFLAG,      obj)) color = warn_color;
+  else if (selected_color  != NULL && TEST_FLAG (SELECTEDFLAG,  obj)) color = selected_color;
+  else if (connected_color != NULL && TEST_FLAG (CONNECTEDFLAG, obj)) color = connected_color;
+  else if (found_color     != NULL && TEST_FLAG (FOUNDFLAG,     obj)) color = found_color;
+  else                                                                color = normal_color;
 
   gui->graphics->set_color (Output.fgGC, color);
 }
@@ -103,7 +102,7 @@ set_object_color (AnyObjectType *obj,
 static void
 set_layer_object_color (LayerType *layer, AnyObjectType *obj)
 {
-  set_object_color (obj, NULL, layer->SelectedColor, PCB->ConnectedColor, layer->Color);
+  set_object_color (obj, NULL, layer->SelectedColor, PCB->ConnectedColor, PCB->FoundColor, layer->Color);
 }
 
 /*---------------------------------------------------------------------------
@@ -204,7 +203,7 @@ draw_pin (PinType *pin, bool draw_hole)
   else
     set_object_color ((AnyObjectType *)pin,
                       PCB->WarnColor, PCB->PinSelectedColor,
-                      PCB->ConnectedColor, PCB->PinColor);
+                      PCB->ConnectedColor, PCB->FoundColor, PCB->PinColor);
 
   _draw_pv (pin, draw_hole);
 }
@@ -224,7 +223,7 @@ draw_via (PinType *via, bool draw_hole)
   else
     set_object_color ((AnyObjectType *)via,
                       PCB->WarnColor, PCB->ViaSelectedColor,
-                      PCB->ConnectedColor, PCB->ViaColor);
+                      PCB->ConnectedColor, PCB->FoundColor, PCB->ViaColor);
 
   _draw_pv (via, draw_hole);
 }
@@ -298,7 +297,7 @@ draw_pad (PadType *pad)
     gui->graphics->set_color (Output.fgGC, PCB->PinColor);
   else
     set_object_color ((AnyObjectType *)pad, PCB->WarnColor,
-                      PCB->PinSelectedColor, PCB->ConnectedColor,
+                      PCB->PinSelectedColor, PCB->ConnectedColor, PCB->FoundColor,
                       FRONT (pad) ? PCB->PinColor : PCB->InvisibleObjectsColor);
 
   _draw_pad (Output.fgGC, pad, false, false);
@@ -403,7 +402,7 @@ hole_callback (const BoxType * b, void *cl)
     {
       set_object_color ((AnyObjectType *) pv,
                         PCB->WarnColor, PCB->ViaSelectedColor,
-                        NULL, Settings.BlackColor);
+                        NULL, NULL, Settings.BlackColor);
 
       gui->graphics->set_line_cap (Output.fgGC, Round_Cap);
       gui->graphics->set_line_width (Output.fgGC, 0);
@@ -444,7 +443,7 @@ rat_callback (const BoxType * b, void *cl)
   RatType *rat = (RatType *)b;
 
   set_object_color ((AnyObjectType *) rat, NULL, PCB->RatSelectedColor,
-                    PCB->ConnectedColor, PCB->RatColor);
+                    PCB->ConnectedColor, PCB->FoundColor, PCB->RatColor);
 
   if (Settings.RatThickness < 100)
     rat->Thickness = pixel_slop * Settings.RatThickness;
diff --git a/src/global.h b/src/global.h
index 7b966dc..91cb74a 100644
--- a/src/global.h
+++ b/src/global.h
@@ -505,7 +505,7 @@ typedef struct PCBType
     *InvisibleObjectsColor,
     *InvisibleMarkColor,
     *ElementSelectedColor,
-    *RatSelectedColor, *ConnectedColor, *WarnColor, *MaskColor;
+    *RatSelectedColor, *ConnectedColor, *FoundColor, *WarnColor, *MaskColor;
   long CursorX,			/* cursor position as saved with layout */
     CursorY, Clipping;
   Coord Bloat,			/* drc sizes saved with layout */
@@ -636,6 +636,7 @@ typedef struct			/* some resources... */
     *ElementSelectedColor,
     *RatSelectedColor,
     *ConnectedColor,
+    *FoundColor,
     *OffLimitColor,
     *GridColor,
     *LayerColor[MAX_LAYER],
diff --git a/src/gpcb-menu.res.in b/src/gpcb-menu.res.in
index f28edff..47bee40 100644
--- a/src/gpcb-menu.res.in
+++ b/src/gpcb-menu.res.in
@@ -233,9 +233,11 @@ MainMenu =
 #
   {"Select" m=l
    {"Select all visible" Select(All)}
+   {"Select all found" Select(Found)}
    {"Select all connected" Select(Connection)}
    -
    {"Unselect all" Unselect(All)}
+   {"Unselect all found" Unselect(Found)}
    {"Unselect all connected" Unselect(Connection)}
    -
    {"Select by name"
diff --git a/src/main.c b/src/main.c
index 3ab4d56..1939b01 100644
--- a/src/main.c
+++ b/src/main.c
@@ -765,7 +765,17 @@ Color to indicate connections. Default: @samp{#00ff00}
 %end-doc
 */
   COLOR (ConnectedColor, "#00ff00", "connected-color",
-	 "color to indicate connections"),
+	 "color to indicate physically connected objects"),
+
+/* %start-doc options "3 Colors"
+@ftable @code
+@item --connected-color <string>
+Color to indicate connections. Default: @samp{#00ff00}
+@end ftable
+%end-doc
+*/
+  COLOR (FoundColor, "#ff00ff", "found-color",
+	 "color to indicate logically connected objects"),
 
 /* %start-doc options "3 Colors"
 @ftable @code
diff --git a/src/pcb-menu.res.in b/src/pcb-menu.res.in
index 84c4a45..e37bd7c 100644
--- a/src/pcb-menu.res.in
+++ b/src/pcb-menu.res.in
@@ -219,9 +219,11 @@ MainMenu =
 
   {Select
    {"Select all visible objects" Select(All)}
+   {"Select all found objects" Select(Found)}
    {"Select all connected objects" Select(Connection)}
    -
    {"Unselect all objects" Unselect(All)}
+   {"unselect all found objects" Unselect(Found)}
    {"unselect all connected objects" Unselect(Connection)}
    -
    {"Select by name" foreground=grey50 sensitive=false}
diff --git a/src/select.h b/src/select.h
index 3e0369f..651ad00 100644
--- a/src/select.h
+++ b/src/select.h
@@ -40,6 +40,7 @@ bool SelectObject (void);
 bool SelectBlock (BoxType *, bool);
 bool SelectedOperation (ObjectFunctionType *, bool, int);
 void *ObjectOperation (ObjectFunctionType *, int, void *, void *, void *);
+bool SelectFound (bool);
 bool SelectConnection (bool);
 
 #if defined(HAVE_REGCOMP) || defined(HAVE_RE_COMP)
diff --git a/src/strflags.c b/src/strflags.c
index 8c2cecb..f29da5b 100644
--- a/src/strflags.c
+++ b/src/strflags.c
@@ -109,7 +109,7 @@ static FlagBitsType object_flagbits[] = {
   { EDGE2FLAG, N ("edge2"), ALL_TYPES },
   { FULLPOLYFLAG, N ("fullpoly"), POLYGON_TYPE},
   { NOPASTEFLAG, N ("nopaste"), PAD_TYPE },
-  { RATFOUNDFLAG, N ("ratfound"), ALL_TYPES }
+  { CONNECTEDFLAG, N ("connected"), ALL_TYPES }
 };
 
 static FlagBitsType pcb_flagbits[] = {
diff --git a/src/undo.h b/src/undo.h
index f59c6fe..c52b051 100644
--- a/src/undo.h
+++ b/src/undo.h
@@ -33,7 +33,7 @@
 #include "global.h"
 
 #define DRAW_FLAGS	(RATFLAG | SELECTEDFLAG | SQUAREFLAG \
-			| HIDENAMEFLAG | HOLEFLAG | OCTAGONFLAG | FOUNDFLAG | RATFOUNDFLAG | CLEARLINEFLAG)
+			| HIDENAMEFLAG | HOLEFLAG | OCTAGONFLAG | CONNECTEDFLAG | FOUNDFLAG | CLEARLINEFLAG)
 
 											/* different layers */
